cpp/src/Slice/Parser.cpp

```diff
@@ -5268,72 +5260,76 @@ Slice::Enum::recDependencies(set<ConstructedPtr>&)
     // An Enum does not have any dependencies.
 }
 
-Slice::Enum::Enum(const ContainerPtr& container, const string& name) :
+Slice::Enum::Enum(const ContainerPtr& container, const string& name, const BuiltinPtr& underlying) :
     SyntaxTreeBase(container->unit()),
     Container(container->unit()),
     Type(container->unit()),
     Contained(container, name),
     Constructed(container, name),
+    _underlying(underlying), // can be null
     _explicitValue(false),
-    _minValue(Int32Max),
-    _maxValue(0),
+    _minValue(INT64_MAX),
+    _maxValue(INT64_MIN),
     _lastValue(-1)
 {
 }
 
-int
+int64_t
 Slice::Enum::newEnumerator(const EnumeratorPtr& p)
 {
-    if(p->explicitValue())
+    int64_t rangeMin = 0;
+    uint64_t rangeMax = INT32_MAX;
+    if (_underlying)
     {
-        _explicitValue = true;
-        _lastValue = p->value();
+        tie(rangeMin, rangeMax) = _underlying->integralRange();
+    }
 
-        if(_lastValue < 0)
-        {
-            string msg = "value for enumerator `" + p->name() + "' is out of range";
-            _unit->error(msg);
-        }
+    if (p->explicitValue())
+    {
+        _explicitValue = true;
+        _lastValue  = p->value();
     }
     else
     {
-        if(_lastValue == Int32Max)
+        _lastValue++;
+    }
+
+    if (_lastValue < rangeMin || _lastValue > static_cast<int64_t>(rangeMax))
+    {
+        ostringstream oss;
+        oss.imbue(underscoreSeparatorLocale);
+        oss << "value " << _lastValue << " for enumerator `" << p->name() << "'";
+        if (_underlying)
         {
-            string msg = "value for enumerator `" + p->name() + "' is out of range";
-            _unit->error(msg);
+            oss << " is outside type's " << _underlying->kindAsString() << " range: [" << rangeMin << ".." << rangeMax
```

Comment:
Same as the other comment, but even if you decide to not change it, there's a typo here:
`type's " << _underlying->kindAsString() << " range`
Should be
`type " << _underlying->kindAsString() << "'s range`

Reply:
Fixed.